Improve performance of decoding

taiki-e opened the pull request:
This change makes decoding about 10x faster than the current implementation, on my machine.

Before:
```
hex_decode              time:   [128.28 us 129.41 us 130.70 us]
```
After:
```
hex_decode              time:   [12.206 us 12.242 us 12.284 us]                        
```

<details>

<summary> Benchmarks per commit</summary>

Current main branch (aa8f3001fed80718ea8e33ab5e278351d3ef5036)
```
hex_decode              time:   [128.28 us 129.41 us 130.70 us]
```

First commit of this PR: Use lookup table for decoding (870590e2284250c4c8789955f7664d33dc15a2ae)
```
hex_decode              time:   [63.538 us 63.936 us 64.419 us]
                        change: [-50.833% -50.462% -50.088%] (p = 0.00 < 0.05)
                        Performance has improved.
```
Second commit of this PR: Inline val function (3b75b32a17153dc467fc403bbf60bf99fb9c0a72)
```
hex_decode              time:   [32.263 us 32.382 us 32.521 us]
                        change: [-49.441% -49.098% -48.682%] (p = 0.00 < 0.05)
                        Performance has improved.
```

Third commit of this PR: Use chunks_exact instead of chunks (ff1d11514e0eb7116aca65f8f69d58a061ee2382)

```
hex_decode              time:   [24.701 us 24.814 us 24.936 us]
                        change: [-24.459% -23.911% -23.395%] (p = 0.00 < 0.05)
                        Performance has improved.
```

Fourth commit of this PR: Use decode_to_slice in Vec::from_hex (4fd0ed8d81c24785c44a33a54f7ccfeb94d13999)

```
hex_decode              time:   [16.060 us 16.297 us 16.608 us]
                        change: [-35.137% -34.468% -33.556%] (p = 0.00 < 0.05)
                        Performance has improved.
```

Fifth commit of this PR: Inline decode_to_slice function (3e2aa25cd58c6aca9aefec7c40cc48aa7c2b118b)

```
hex_decode              time:   [12.206 us 12.242 us 12.284 us]                        
                        change: [-21.954% -20.254% -18.150%] (p = 0.00 < 0.05)
                        Performance has improved.
```

</details>

taiki-e committed:
Use lookup table for decoding

taiki-e committed:
Inline val function

taiki-e committed:
Use chunks_exact instead of chunks

taiki-e committed:
Use decode_to_slice in Vec::from_hex

taiki-e committed:
Inline decode_to_slice function

taiki-e committed:
Fix build errors

Luro02 reviewed (changes requested):
Looks good :)

taiki-e commented on src/lib.rs:
```diff
@@ -40,9 +40,9 @@
+use alloc::{string::String, vec, vec::Vec};
 
-use core::iter;
+use core::{iter, u8};
```
I used u8 module because the u8::MAX associated constant requires a relatively new compiler.

See also: https://github.com/KokaKiwi/rust-hex/issues/55

Luro02 replied:
You have a point. Personally, I do not really care about MSRV (because I want to use the new features and updating the compiler is just a simple `rustup update`) so I did not think much about it.

https://github.com/KokaKiwi/rust-hex/pull/58#issuecomment-860209726

taiki-e commented on src/lib.rs:
```diff
@@ -340,7 +371,7 @@ fn generate_iter(len: usize) -> impl Iterator<Item = (usize, usize)> {
 #[inline]
 #[must_use]
-const fn byte2hex(byte: u8, table: &[u8; 16]) -> (u8, u8) {
+fn byte2hex(byte: u8, table: &[u8; 16]) -> (u8, u8) {
```
byte2hex is a private function, so I don't think this is a breaking change.

Luro02 replied:
oh you are right, my bad ._.

taiki-e commented on src/lib.rs:
```diff
@@ -172,16 +172,47 @@ pub trait FromHex: Sized {
+
+// Lookup table for ascii to hex decoding.
+#[rustfmt::skip]
+static DECODE_TABLE: [u8; 256] = [
```
The lookup table is relatively large, so using const may bloat the binary size by inlining.

IIRC, the standard library uses static for that use-case: https://github.com/rust-lang/rust/blob/1.54.0/library/core/src/unicode/unicode_data.rs

See also https://stackoverflow.com/questions/52751597/what-is-the-difference-between-a-constant-and-a-static-variable-and-which-should

taiki-e replied:
By the way, benchmarking (on my machine) shows that changing this to const does not change the performance.

taiki-e commented on src/lib.rs:
```diff
@@ -172,16 +172,47 @@ pub trait FromHex: Sized {
+];
+
+#[inline]
+fn val(bytes: &[u8], idx: usize) -> Result<u8, FromHexError> {
```
The new implementation of val function cannot be const

Luro02 replied:
If the table were `const` and not `static`, this function could be constant: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&code=const%20__%3A%20u8%20%3D%20u8%3A%3AMAX%3B%0D%0A%23%5Bderive(Debug%2C%20Clone%2C%20Copy%2C%20PartialEq)%5D%0D%0Apub%20enum%20FromHexError%20%7B%0D%0A%20%20%20%20InvalidHexCharacter%20%7B%20c%3A%20char%2C%20index%3A%20usize%20%7D%2C%0D%0A%20%20%20%20OddLength%2C%0D%0A%20%20%20%20InvalidStringLength%2C%0D%0A%7D%0D%0A%2F%2F%20Lookup%20table%20for%20ascii%20to%20hex%20decoding.%0D%0A%23%5Brustfmt%3A%3Askip%5D%0D%0Aconst%20DECODE_TABLE%3A%20%5Bu8%3B%20256%5D%20%3D%20%5B%0D%0A%20%20%20%20%2F%2F%20%20%201%20%20%202%20%20%203%20%20%204%20%20%205%20%20%206%20%20%207%20%20%208%20%20%209%20%20%20a%20%20%20b%20%20%20c%20%20%20d%20%20%20e%20%20%20f%0D%0A%20%20%20%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20%2F%2F%200%0D%0A%20%20%20%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20%2F%2F%201%0D%0A%20%20%20%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20%2F%2F%202%0D%0A%20%20%20%20%200%2C%20%201%2C%20%202%2C%20%203%2C%20%204%2C%20%205%2C%20%206%2C%20%207%2C%20%208%2C%20%209%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20%2F%2F%203%0D%0A%20%20%20%20__%2C%2010%2C%2011%2C%2012%2C%2013%2C%2014%2C%2015%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20%2F%2F%204%0D%0A%20%20%20%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20%2F%2F%205%0D%0A%20%20%20%20__%2C%2010%2C%2011%2C%2012%2C%2013%2C%2014%2C%2015%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20%2F%2F%206%0D%0A%20%20%20%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20%2F%2F%207%0D%0A%20%20%20%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20%2F%2F%208%0D%0A%20%20%20%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20%2F%2F%209%0D%0A%20%20%20%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20%2F%2F%20a%0D%0A%20%20%20%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20%2F%2F%20b%0D%0A%20%20%20%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20%2F%2F%20c%0D%0A%20%20%20%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20%2F%2F%20d%0D%0A%20%20%20%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20%2F%2F%20e%0D%0A%20%20%20%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20__%2C%20%2F%2F%20f%0D%0A%5D%3B%0D%0A%0D%0A%23%5Binline%5D%0D%0Aconst%20fn%20val(bytes%3A%20%26%5Bu8%5D%2C%20idx%3A%20usize)%20-%3E%20Result%3Cu8%2C%20FromHexError%3E%20%7B%0D%0A%20%20%20%20let%20upper%20%3D%20DECODE_TABLE%5Bbytes%5B0%5D%20as%20usize%5D%3B%0D%0A%20%20%20%20let%20lower%20%3D%20DECODE_TABLE%5Bbytes%5B1%5D%20as%20usize%5D%3B%0D%0A%20%20%20%20if%20upper%20%3D%3D%20u8%3A%3AMAX%20%7B%0D%0A%20%20%20%20%20%20%20%20return%20Err(FromHexError%3A%3AInvalidHexCharacter%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20c%3A%20bytes%5B0%5D%20as%20char%2C%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20index%3A%20idx%2C%0D%0A%20%20%20%20%20%20%20%20%7D)%3B%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20if%20lower%20%3D%3D%20u8%3A%3AMAX%20%7B%0D%0A%20%20%20%20%20%20%20%20return%20Err(FromHexError%3A%3AInvalidHexCharacter%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20c%3A%20bytes%5B1%5D%20as%20char%2C%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20index%3A%20idx%20%2B%201%2C%0D%0A%20%20%20%20%20%20%20%20%7D)%3B%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20Ok((upper%20%3C%3C%204)%20%7C%20lower)%0D%0A%7D%0D%0A%0D%0Afn%20main()%20%7B%7D)

KokaKiwi committed:
Merge branch 'main' into pref

KokaKiwi committed:
Merge branch 'pref' of github.com:taiki-e/rust-hex into taiki-e-pref

KokaKiwi committed:
refactor: improve code style and clarity in src/lib.rs

- Remove unused `u8` import from `core` module
- Add lint allow annotation for `__` constant to suppress non_upper_case_globals warning
- Refactor `decode_in_slice()` to use slice-based `val()` function for cleaner implementation

KokaKiwi committed:
Apply suggestions from code review

Co-authored-by: Lucas <24826124+Luro02@users.noreply.github.com>

KokaKiwi committed:
Merge branch 'pref' of github.com:taiki-e/rust-hex into taiki-e-pref
